Control logging via command line or environment variable

foresto opened the pull request:
Nheko is very chatty in its log output, generating log noise (which complicates diagnostics) and needless disk writes (which affect power consumption and SSD life).  This patch provides control over it, with command line options for:

- Setting the global log level
- Setting log levels for individual components (loggers)
- Disabling log file writes
- Disabling stderr writes (which are often written to ~/.xsession-errors)

The old --debug command line option still works, at least for now, and is overridden by the new fine-grained option.

Log levels and output type (file/stderr) can also be set by environment variables. (This is especially convenient for flatpak users, who can `flatpak override --env=` instead of having to copy and edit the .desktop file to add command line options.)

Partially addresses #665.

foresto commented:
Looks like lint flagged a few formatting issues. I can easily reorder the #include lines, but the rest doesn't make sense to me.

Why does it want to indent with 2 spaces when the existing code is clearly using 4 spaces? Was lint misconfigured?

Why does it want to insert more than 30 spaces of indentation in the last flagged section of main.cpp? That wastes a lot of space and doesn't align with anything.

foresto commented:
What version of lint is being used here? What configuration?

redsky17 commented:
we use clang format for linting and our configuration is in the root of the repository under `.clang-format.`  Our lint isn't misconfigured.  I promise it's consistent.  If it wasn't, the 'lint' step would mark everything inconsistent as needing to be fixed.  You should be able to run `make lint` from the root nheko git directory and it'll fix everything fo ryou.

foresto commented:
Ah, there it is. 

> ContinuationIndentWidth: 2
> IndentWidth:     4

That's what I mean by inconsistent. (Maybe I should have said "internally inconsistent".) If that's the format you prefer, though, that's what I'll use. :)

Thanks for the info. I'll reformat and resubmit.

deepbluev7 commented:
Can you update the manpage as well? I think that just mentions the debug switch atm.

deepbluev7 commented on src/main.cpp:
```diff
@@ -202,8 +202,20 @@ main(int argc, char *argv[])
+    QCommandLineOption debugOption(QStringLiteral("debug"),
+                                   QStringLiteral("Alias for \"--loglevel trace\"."));
     parser.addOption(debugOption);
+    QCommandLineOption noLog(QStringLiteral("nolog"), QStringLiteral("No log file output."));
+    parser.addOption(noLog);
+    QCommandLineOption noStderr(QStringLiteral("nostderr"), QStringLiteral("No stderr output."));
+    parser.addOption(noStderr);
```
How about instead of making this 2 options, we make it a list, i.e.: "--log-output=file,stderr"?

foresto replied:
Hm...

What's your goal with that approach? 

I thought at first that it might make the --help text slightly shorter, but on further consideration, I suspect that describing the valid arguments would end up taking two lines anyway.

It would also make command lines longer.

It would also require a string tokenizer to parse the commas. (We get this for free in the loglevel argument because spdlog handles it internally.) I was hoping to keep this simple, minimizing the added code and the time I spend on it.

C's strtok() would do the job, but would come with the risks of using C-style strings in the code base. I'm not aware of a standard C++ equivalent. QStringTokenizer() doesn't exist in Qt 5. I guess QString::split() would be the way to go if I were to write a tokenizer.

How important is this to you?

foresto replied:
My vote is -1 on this suggestion, mainly for usability reasons.

The deciding factor for me is that it's much easier to type --nolog than --log-output=stderr every time I run nheko from the command line, which I do often when testing new code.

Writing a shell completion script wouldn't solve that, since completion would require two steps: --l<TAB>s<TAB> .

deepbluev7 replied:
Well, splitting a string in Qt is just str.split(','), which doesn't sound like that much effort. My issue is mostly that the parameter looks weird, so I guess it is cosmetical. Doing --log-no-stderr might be an acceptable alternative as well or something like that? I would assume you use completion anyway to just run your previous command.

foresto replied:
> I would assume you use completion anyway to just run your previous command.

If that were the case, I wouldn't be spending time arguing the point. ;)

Command history can somewhat gloss over an annoying command line option, but only if I'm running exactly the same command line every time. When I have to edit it or iterate through the history to find the right one, it doesn't help. (I've been doing this a lot lately while comparing palettes, for example.)

Tab completion can sometimes help with annoying command line options, but only if they have unique prefixes, and only if a completion script exists (it currently doesn't) for the shell I'm in (I don't always use bash).

> My issue is mostly that the parameter looks weird, so I guess it is cosmetical. Doing --log-no-stderr might be an acceptable alternative as well or something like that?

I see. I guess the divide here is that you're more concerned with the appearance, while I'm more concerned with ease of use. (And just to be clear, I do value good aesthetics, so I sympathize with your position. :)

I don't love --log-no-stderr, but maybe we can still find some middle ground.

How about --stderr-off and --log-off? That approach has the word separation that you like, without being terribly long. The word "off" also matches spdlog's log level name, making it easy to remember. The dashes between words would still be annoying to type, but the unique prefixes would enable tab completion in any shells for which someone writes completion scripts.

deepbluev7 replied:
Well, log-off is confusing. That turns of stdout or the logfile? And stderr-off just looks weird to me for something that is supposed to affect logs! I don't type different Nheko commands that often, but I need to explain stuff a hundred times over, so self explanatory stuff is pretty important to me! :D

foresto replied:
> Well, log-off is confusing. That turns of stdout or the logfile?
> And stderr-off just looks weird to me for something that is supposed to affect logs!

Hm... It didn't occur to me that they could confuse a user, because --stderr-off and --log-off are explicitly named after the thing they affect, and because I don't think of stderr or stdout as logs at all. (They only become logs if someone redirects them to a file, but I guess the fact that we use a logging system internally to generate their output might make it easy for programmers to think of them as logs.)

Okay, I'll think about alternatives.

foresto replied:
I've been thinking about ways to refine your original idea to be less burdensome to type. So, instead of `--log-output=file,stderr`, how about this:

1. Starting with a concise option name would help a lot by itself: `--log=file,stderr`
2. Accepting abbreviated values would further reduce typing overhead, so these would work as well: `--log=std` or even `--log=f,s` (The help text would show the full value names, of course.)
3. An argument starting with `/` could be treated as a custom log file location: `--log=/tmp/debug.log` (This is somewhat common already in the unix world.)
4. The option could be accepted multiple times: `--log=/tmp/debug.log --log=stderr`
5. To disable output entirely: `--log=none` or maybe `--log=off` if we want to adopt spdlog's keyword.

Of course, implementing this would be more work for me, but if it yields something that makes both of us happy, I'm willing to do it.

deepbluev7 replied:
I think I would leave out the abbreviated options and treating stuff with a / as a filename might be a bit too much magic, since that won't work with logfiles in cwd or on Windows. Also having 3 options to control log verbosity seems like a bit much. But I guess --log=file,stderr is fine, although I still prefer a common prefix like --log-to=file.

In the end I think it doesn't make sense to make this too featurefull. Almost nobody will be using it and I don't want this to ever break.

foresto replied:
Windows paths: Yes, and drive letters as well. Relative paths like those in cwd: Yep. Part of the "more work for me" would have been handling those issues, of course.

Your point about too much magic is well taken, though. Consider those ideas abandoned. :) Expect a new draft soon.

deepbluev7 commented on src/main.cpp:
```diff
@@ -202,8 +202,20 @@ main(int argc, char *argv[])
+      QStringLiteral("loglevel"),
+      QStringLiteral("Set the log level, or a <component>=<level> list like \"warn,ui=off\". "
+                     "levels:{trace,debug,info,warning,error,critical,off} "
+                     "components:{crypto,db,net,qml,ui}"),
```
There is actually one more, "mtx" or so from the mtxclient lib

foresto replied:
Thanks. I've added it in my local copy & will include it with my next update.

foresto replied:
Actually, from a quick look at the mtxclient code, I don't think it adds its logger to [spdlog's registry](https://github.com/gabime/spdlog/wiki/5.-Logger-registry#registering-new-loggers). Did I miss something there? If not, there's no point in mentioning "mtx" in our help text, since spdlog will ignore it.

foresto replied:
I'll try registering mtxclient's logger from within Nheko. I was concerned at first about what would happen if mtxclient starts registering its own logger in some future version, but it turns out there's a [documented exception](https://github.com/gabime/spdlog/wiki/5.-Logger-registry#registry-conflicts) for that, so I think I can handle it safely.

deepbluev7 replied:
I didn't know about registering loggers. I think we should just not register the mtxclient logger in Nheko and just change that in mtxclient to always register it. But it makes sense to already document it here, imo.

foresto replied:
Whose project is mtxclient? Is this change something should be done upstream?

deepbluev7 replied:
https://github.com/Nheko-Reborn/mtxclient is the upstream

foresto replied:
Oh! Interesting. I thought most Matrix clients used a library from matrix.org, or else a stand-alone project. I didn't expect Nheko to have a custom one.

deepbluev7 commented on src/main.cpp:
```diff
@@ -202,8 +202,20 @@ main(int argc, char *argv[])
     parser.addVersionOption();
-    QCommandLineOption debugOption(QStringLiteral("debug"), QStringLiteral("Enable debug output"));
+    QCommandLineOption logLevel(
+      QStringLiteral("loglevel"),
```
Nitpick, but can we name this "log-level"? At least if we go with "log-output", then it would at least be symmetric.

foresto replied:
I like symmetry. I'm reluctant to add dashes in the middle of either option, though, because they make typing awkward. For symmetry's sake, I'll leave this one alone until we reach a decision on the other one.

deepbluev7 commented on src/main.cpp:
```diff
@@ -202,8 +202,20 @@ main(int argc, char *argv[])
-    QCommandLineOption debugOption(QStringLiteral("debug"), QStringLiteral("Enable debug output"));
+    QCommandLineOption logLevel(
+      QStringLiteral("loglevel"),
+      QStringLiteral("Set the log level, or a <component>=<level> list like \"warn,ui=off\". "
```
```suggestion
      QStringLiteral("Specify the log levels. You can set the global log level or the log level for individual components or a combination of both. For example to set the default log level to warn but disable logging for the ui component, pass '--log-level=warn,ui=off'."
```

foresto replied:
Thanks. We need a trailing space at the end of that sentence, though. Does github let you edit your diff? If not, I can make the change locally and include it in my next update.

deepbluev7 commented on src/main.cpp:
```diff
@@ -254,13 +266,13 @@ main(int argc, char *argv[])
-        nhlog::init(QStringLiteral("%1/nheko.log")
-                      .arg(QStandardPaths::writableLocation(QStandardPaths::CacheLocation))
-                      .toStdString());
+        std::string level =
+          (parser.isSet(debugOption) ? "trace," : "") + parser.value(logLevel).toStdString();
```
I think the loglevel option should overwrite the debug option, i.e.:

```suggestion
        std::string level =
          parser.isSet(loglevel) ? parser.value(logLevel).toStdString() : (parser.isSet(debugOption) ? "trace" : ""));
```

foresto replied:
Levels set by --loglevel already override the --debug option. For example, `--debug --loglevel db=error` enables verbose logging for everything except db.

Or do you want --debug to be ignored in cases where someone uses both?

foresto replied:
We might also consider deprecating or removing the --debug option, since --loglevel offers a superset of its functionality. I kept it for backward compatibility, since I don't know who is using it.

deepbluev7 replied:
I think it should completely override the debug option. Otherwise it would be too confusing for me. I would not remove the debug option for now, since it is a bit of a pain to figure out what version people use before asking them for logs.

foresto replied:
Okay, then. I'll do that.

foresto commented:
- Rephrased the help text to be more descriptive, following your suggested change, and to be more specific about the syntax.
- Updated the man page.
- Added support for the mtx logger.

foresto commented:
Changes since last draft:

- `--log-level` instead of `--loglevel`.
- `--log=<type>` instead of `--nolog`, `--nostderr`.
- `NHEKO_LOG` environment variable. (The comma-separated `--log=file,stderr` syntax was superfluous before, since it was the default, but it can now be useful for overriding the environment variable.)
- Translatable command line help text.

> I guess --log=file,stderr is fine, although I still prefer a common prefix like --log-to=file.

I went with --log because easy typing is important to me. Thanks for allowing it. However:

What about satisfying the desire for a common prefix while still allowing easy typing, by adding short forms of these options? For example:

`-l` `--log-level`
`-L` `--log-to` (or `--log-type` or `--log-output`)

(I'm a little embarrassed for not suggesting this common pattern earlier. I usually reserve short command line options until I know what the most common operations will be, since there are only so many letters available. Since Nheko is already in real-world use and its common operations are all through the GUI, this isn't really an issue here.)

foresto commented:
- Rebased to master.
- Implemented short & long options. (Both long options now use the `--log-` prefix).
- Updated environment variable to match its corresponding renamed option.
- Alphabetically ordered option help text.

deepbluev7 commented on src/main.cpp:
```diff
@@ -254,15 +270,35 @@ main(int argc, char *argv[])
-        nhlog::init(QStringLiteral("%1/nheko.log")
-                      .arg(QStandardPaths::writableLocation(QStandardPaths::CacheLocation))
-                      .toStdString());
+        QStringList levels;
+        levels.append(qEnvironmentVariable("NHEKO_LOG_LEVEL"));
+        if (parser.isSet(logLevel)) {
+            levels.append(parser.value(logLevel));
+        } else if (parser.isSet(debugOption)) {
+            levels.append("trace");
+        }
```
I think I would still prefer overwrite semantics, i.e. if --log-level is set, use that, else if --debug is set, use that, else use env flag.

foresto replied:
Oh! When you mentioned overwriting the debug option earlier, I didn't assume you wanted to overwrite the environment variable as well. Okay, I'll make it do that.

deepbluev7 commented:
I think this looks pretty nice now, I would just prefer it if the different log level flags were not merged, but instead would overwrite each other. Otherwise it can be hard to troubleshoot, why something behaves weirdly, imo.

deepbluev7 commented:
Awesome, thank you for your patience!
